From 4da818c17b6923bbf2c736d3eb2eed063dc475af Mon Sep 17 00:00:00 2001
From: MorsMortium <mors.mortium.8@protonmail.com>
Date: Wed, 19 Aug 2026 15:24:56 +0200
Subject: [PATCH] Multiple changes

Manually hide title buttons in GTK4 too: Fixes Iagno/Reversi extra close button
Delay LibAdwaita/LibHandy header operations until realized: Fixes Eye of GNOME crash
Small string fixes
---
 Source/GTK-NoCSD.c | 118 +++++++++++++++++++++++++--------------------
 1 file changed, 66 insertions(+), 52 deletions(-)

diff --git a/Source/GTK-NoCSD.c b/Source/GTK-NoCSD.c
index 8e0ed1c..d5e0fb9 100644
--- a/Source/GTK-NoCSD.c
+++ b/Source/GTK-NoCSD.c
@@ -774,6 +774,7 @@ void (*o_gtk_im_context_set_cursor_location) (GtkIMContext *,
 void (*o_gtk_im_context_set_client_window) (GtkIMContext *, void *) = NULL;
 gboolean (*o_gtk_window_is_active) (GtkWindow *) = NULL;
 void (*o_gtk_widget_get_preferred_width) (GtkWidget *, gint *, gint *) = NULL;
+gboolean (*o_gtk_widget_get_realized) (GtkWidget *) = NULL;
 
 // This is needed by an unavoidable GTK macros for type registration
 #define gtk_widget_get_type(...) o_gtk_widget_get_type(__VA_ARGS__)
@@ -1169,6 +1170,7 @@ void GTKNoCSDGetReferences(bool GetTypes) {
 			LOAD_SYMBOL(Library, gtk_file_chooser_widget_get_type);
 			LOAD_SYMBOL(Library, gtk_widget_get_visible);
 			LOAD_SYMBOL(Library, gtk_button_get_type);
+			LOAD_SYMBOL(Library, gtk_widget_get_realized);
 		}
 
 		if (GTKNoCSDGTKVersion == 4) {
@@ -1407,6 +1409,29 @@ void GTKNoCSDTitleChanged(GtkLabel *Label, G_GNUC_UNUSED GParamSpec
 	GTKNoCSDLabelChange(Label);
 }
 
+void GTKNoCSDForceInvisible(GtkWidget *Widget, G_GNUC_UNUSED GParamSpec *Spec,
+	G_GNUC_UNUSED gpointer Data) {
+	// No show all does not always help, if visibility is set manually. It is
+	// forced off here
+
+	o_gtk_widget_set_visible(Widget, false);
+	if (o_gtk_widget_set_no_show_all != NULL) {
+		o_gtk_widget_set_no_show_all(Widget, true);
+	}
+}
+
+void GTKNoCSDSetForceInvisible(GtkWidget *Widget) {
+	// Add function to force off visibility, call it once for initial set up
+
+	if (!o_g_signal_handler_find((GObject *) Widget, G_SIGNAL_MATCH_FUNC,
+		0, 0, NULL, (GCallback *) GTKNoCSDForceInvisible, NULL)) {
+		// WARNING: Macro
+		g_signal_connect(Widget, "notify::visible",
+			G_CALLBACK(GTKNoCSDForceInvisible), NULL);
+	}
+	GTKNoCSDForceInvisible(Widget, NULL, NULL);
+}
+
 bool GTKNoCSDCheckGTK4Header(GtkWidget *Widget) {
 	// Check if a widget is a header bar and if it is and the title widget is
 	// empty, disable showing it, plus signal it back
@@ -1422,6 +1447,11 @@ bool GTKNoCSDCheckGTK4Header(GtkWidget *Widget) {
 		GTKNoCSDTitleChanged((GtkLabel *) Widget, NULL, NULL);
 	}
 
+	// If widget is a close button, hide it
+	if (GTKNoCSDHasClass(Widget, "titlebutton")) {
+		GTKNoCSDSetForceInvisible(Widget);
+	}
+
 	// Force set decoration layouts to empty
 	if (GTKNoCSDAdwHeaderBar((GObject *) Widget)) {
 		o_adw_header_bar_set_decoration_layout((AdwHeaderBar *) Widget, "");
@@ -1460,14 +1490,6 @@ GtkWidget *GTKNoCSDGetGTK4Headers(GtkWidget *Widget, bool GetWidget) {
 	return NULL;
 }
 
-gboolean GTKNoCSDSetGTK4Headers(gpointer Data) {
-	// GTK callback for GTKNoCSDGetGTK4Headers
-
-	GTKNoCSDGetGTK4Headers((GtkWidget *) Data, false);
-
-	return FALSE;
-}
-
 bool GTKNoCSDTest(GtkWidget *Widget) {
 	// Simple test if the widget we deal with is our own
 
@@ -1569,29 +1591,6 @@ void GTKNoCSDGtkBoxAdd(GtkWidget *Box, GtkWidget *Widget) {
 	}
 }
 
-void GTKNoCSDForceInvisible(GtkWidget *Widget, G_GNUC_UNUSED GParamSpec *Spec,
-	G_GNUC_UNUSED gpointer Data) {
-	// No show all does not always help, if visibility is set manually. It is
-	// forced off here
-
-	o_gtk_widget_set_visible(Widget, false);
-	if (o_gtk_widget_set_no_show_all != NULL) {
-		o_gtk_widget_set_no_show_all(Widget, true);
-	}
-}
-
-void GTKNoCSDGTK3SetForceInvisible(GtkWidget *Widget) {
-	// Add function to force off visibility, call it once for initial set up
-
-	if (!o_g_signal_handler_find((GObject *) Widget, G_SIGNAL_MATCH_FUNC,
-		0, 0, NULL, (GCallback *) GTKNoCSDForceInvisible, NULL)) {
-		// WARNING: Macro
-		g_signal_connect(Widget, "notify::visible",
-			G_CALLBACK(GTKNoCSDForceInvisible), NULL);
-	}
-	GTKNoCSDForceInvisible(Widget, NULL, NULL);
-}
-
 void GTKNoCSDGTK3EmptyTitleChanged(GtkWidget *Header,
 	G_GNUC_UNUSED GParamSpec *Spec, G_GNUC_UNUSED gpointer Data) {
 	// Make the headerbar take as much space as it would naturally
@@ -1655,7 +1654,7 @@ bool GTKNoCSDGetGTK3Headers(GtkWidget *Container) {
 	}
 
 	if (GTKNoCSDHasClass(Container, "titlebutton")) {
-		GTKNoCSDGTK3SetForceInvisible(Container);
+		GTKNoCSDSetForceInvisible(Container);
 		return true;
 	}
 
@@ -1672,12 +1671,12 @@ bool GTKNoCSDGetGTK3Headers(GtkWidget *Container) {
 
 		// Go through all children of child too
 		if (GTKNoCSDGetGTK3Headers(Child)) {
-			GTKNoCSDGTK3SetForceInvisible(Child);
+			GTKNoCSDSetForceInvisible(Child);
 			++Hidden;
 		} else {
-			// If close button, remove
+			// If close button, hide
 			if (GTKNoCSDHasClass(Child, "titlebutton")) {
-				GTKNoCSDGTK3SetForceInvisible(Child);
+				GTKNoCSDSetForceInvisible(Child);
 				++Hidden;
 			}
 		}
@@ -1686,21 +1685,27 @@ bool GTKNoCSDGetGTK3Headers(GtkWidget *Container) {
 	}
 	o_g_list_free(Children);
 
-	// If all widgets got removed, remove the container too (title button)
+	// If all widgets got hidden, hide the container too (title button)
 	if (Amount == Hidden && Amount != 0) {
-		GTKNoCSDGTK3SetForceInvisible(Container);
+		GTKNoCSDSetForceInvisible(Container);
 		return true;
 	}
 
 	return false;
 }
 
-gboolean GTKNoCSDSetGTK3Headers(gpointer Data) {
-	// GTK callback for GTKNoCSDGetGTK3Headers
+void GTKNoCSDSetHeaders(GtkWidget *Widget) {
+	// GTK callback for GTKNoCSDGetGTK3Headers and GTKNoCSDGetGTK4Headers
 
-	GTKNoCSDGetGTK3Headers((GtkWidget *) Data);
+	// WARNING: Macro
+	g_signal_handlers_disconnect_by_func(Widget,
+		(GCallback *) GTKNoCSDSetHeaders, NULL);
 
-	return FALSE;
+	if (GTKNoCSDGTKVersion == 3) {
+		GTKNoCSDGetGTK3Headers(Widget);
+	} else {
+		GTKNoCSDGetGTK4Headers(Widget, false);
+	}
 }
 
 // One time setups
@@ -2365,7 +2370,7 @@ int GTKNoCSDMagic(GtkWindow *Window, size_t Options) {
 
 	// Never touch inspector windows
 	const char *WindowName = o_gtk_widget_get_name((GtkWidget *) Window);
-	if (strstr(WindowName, "GtkInspectorWindow") == WindowName) {
+	if (strncmp(WindowName, "GtkInspectorWindow", 18) == 0) {
 		return 3;
 	}
 
@@ -2544,8 +2549,12 @@ int GTKNoCSDMagic(GtkWindow *Window, size_t Options) {
 
 		// Hide the title widget, if it is not a special control. It might be
 		// added later so we check then
-		o_g_timeout_add(1, GTKNoCSDGTKVersion == 4 ?
-			GTKNoCSDSetGTK4Headers : GTKNoCSDSetGTK3Headers, Window);
+		if (o_gtk_widget_get_realized((GtkWidget *) Window)) {
+			GTKNoCSDSetHeaders((GtkWidget *) Window);
+		} else {
+			g_signal_connect(Window, "realize", G_CALLBACK(GTKNoCSDSetHeaders),
+				NULL);
+		}
 
 		if (GTKNoCSDGTKVersion == 3) {
 			// Unset the header for the window, but save it for returning it in
@@ -2715,7 +2724,12 @@ int GTKNoCSDMagic(GtkWindow *Window, size_t Options) {
 			!GTKNoCSDAdwWindow((GObject *) Window)) {
 			// WARNING: Macro
 			if (strcmp("AdwGizmo", o_g_type_name(G_OBJECT_TYPE(Header))) != 0) {
-				o_g_timeout_add(1, GTKNoCSDSetGTK4Headers, Header);
+				if (o_gtk_widget_get_realized(Header)) {
+					GTKNoCSDSetHeaders(Header);
+				} else {
+					g_signal_connect(Header, "realize",
+						G_CALLBACK(GTKNoCSDSetHeaders), NULL);
+				}
 			}
 		}
 
@@ -3477,7 +3491,7 @@ void GTKNoCSDHeaderBarParent(GtkWidget *Widget, G_GNUC_UNUSED GParamSpec *Spec,
 	GtkWidget *Parent = o_gtk_widget_get_parent(Widget);
 	if (!GTKNoCSDCSD && GTKNoCSDGtkWindow((GObject *) Parent) &&
 		GTKNoCSDHasClass(Widget, "default-decoration")) {
-		GTKNoCSDGTK3SetForceInvisible(Widget);
+		GTKNoCSDSetForceInvisible(Widget);
 	}
 }
 
@@ -3882,13 +3896,13 @@ gboolean g_module_symbol(GModule *Module, const gchar *Name, gpointer *Symbol) {
 	// there is a chance of getting what is used by the library. After this also
 	// get the used types. Used by script languages on sandboxed systems (Nix)
 	if (!GTKNoCSDGotPlatform) {
-		if (GTKNoCSDGTKVersion == 4 && strstr(Name,
-			"adw_") == Name && o_adw_window_get_type == NULL) {
+		if (GTKNoCSDGTKVersion == 4 && strncmp(Name, "adw_", 4) == 0 &&
+			o_adw_window_get_type == NULL) {
 			GTKNoCSDGetAdwSymbols(Module, true);
 			GTKNoCSDGetAdwTypes();
 			GTKNoCSDGotPlatform = true;
-		} else if (GTKNoCSDGTKVersion == 3 && strstr(Name,
-			"hdy_") == Name && o_hdy_window_get_type == NULL) {
+		} else if (GTKNoCSDGTKVersion == 3 && strncmp(Name, "hdy_", 4) == 0 &&
+			o_hdy_window_get_type == NULL) {
 			GTKNoCSDGetHdySymbols(Module, true);
 			GTKNoCSDGetHdyTypes();
 			GTKNoCSDGotPlatform = true;
@@ -3927,7 +3941,7 @@ gboolean g_module_symbol(GModule *Module, const gchar *Name, gpointer *Symbol) {
 			GET_SYMBOL1(gtk_im_context_set_cursor_location);
 		}
 
-		if (strncmp(Name, "adw_", 0) == 0) {
+		if (strncmp(Name, "adw_", 4) == 0) {
 			GET_SYMBOL1(adw_header_bar_set_title_widget);
 			GET_SYMBOL1(adw_dialog_present);
 			GET_SYMBOL1(adw_application_window_get_visible_dialog);
@@ -3986,7 +4000,7 @@ void *dlsym(void *Handle, const char *Name) {
 		GET_SYMBOL2(gtk_im_context_set_cursor_location);
 	}
 
-	if (strncmp(Name, "adw_", 0) == 0) {
+	if (strncmp(Name, "adw_", 4) == 0) {
 		GET_SYMBOL2(adw_header_bar_set_title_widget);
 		GET_SYMBOL2(adw_dialog_present);
 		GET_SYMBOL2(adw_application_window_get_visible_dialog);
